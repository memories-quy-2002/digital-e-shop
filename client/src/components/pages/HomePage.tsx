import { useContext, useEffect, useState } from "react";
import { IoArrowForward } from "react-icons/io5";
import { Link, useNavigate } from "react-router-dom";
import Cookies from "universal-cookie";
import axios from "../../api/axios";
import iphone from "../../assets/images/iphone.jpg";
import { useToast } from "../../context/ToastContext";
import { UserContext } from "../../context/UserDataContext";
import "../../styles/HomePage.scss";
import { Product } from "../../utils/interface";
import recommendations from "../../utils/recommendations.json";
import NavigationBar from "../common/NavigationBar";
import ProductItem from "../common/ProductItem";
import Layout from "../layout/Layout";

const cookies = new Cookies();
const DISPLAYED_NUMBER = 12;
interface Wishlist {
    id: number;
    product: Product;
}

const HomePage = () => {
    const navigate = useNavigate();
    const [products, setProducts] = useState<Product[]>([]);
    const [wishlist, setWishlist] = useState<Wishlist[]>([]);
    const [userRecommendations, setUserRecommendations] = useState<number[]>(
        []
    );
    const uid =
        cookies.get("rememberMe")?.uid ||
        (sessionStorage["rememberMe"]
            ? JSON.parse(sessionStorage["rememberMe"]).uid
            : "");
    const { userData, loading, fetchUserData } = useContext(UserContext);
    const { addToast } = useToast();

    const handleAddingWishlist = async (
        user_id: string,
        product_id: number
    ) => {
        if (uid === "") {
            addToast(
                "Login required",
                "You need to login to use this feature."
            );
            return;
        }
        try {
            if (wishlist.some((item) => item.product.id === product_id)) {
                const response = await axios.post(`/api/wishlist/delete/`, {
                    uid: user_id,
                    pid: product_id,
                });
                if (response.status === 200) {
                    console.log(response.data.msg);
                    setWishlist((list) =>
                        list.filter((item) => item.product.id !== product_id)
                    );
                    addToast(
                        "Remove wishlist item",
                        "Item removed from wishlist successfully"
                    );
                }
            } else {
                const response = await axios.post("/api/wishlist/", {
                    uid: user_id,
                    pid: product_id,
                });
                if (response.status === 200) {
                    console.log(response.data.msg);
                    const newProduct = products.filter(
                        (product) => product.id === product_id
                    )[0];
                    setWishlist((list) => [
                        ...list,
                        {
                            id: product_id,
                            product: newProduct,
                        },
                    ]);
                    addToast(
                        "Add wishlist item",
                        "Item added to wishlist successfully"
                    );
                }
            }
        } catch (err) {
            console.error(err);
        }
    };
    const handleAddingCart = async (user_id: string, product_id: number) => {
        if (uid === "") {
            addToast(
                "Login required",
                "You need to login to use this feature."
            );
            return;
        }
        try {
            const response = await axios.post("/api/cart/", {
                uid: user_id,
                pid: product_id,
                quantity: 1,
            });
            if (response.status === 200) {
                console.log(response.data.msg);
                addToast("Add cart item", "Product added to cart successfully");
            }
        } catch (err) {
            console.error(err);
        }
    };
    useEffect(() => {
        fetchUserData(uid);
        return () => {};
    }, [uid]);

    useEffect(() => {
        const fetchProducts = async () => {
            try {
                const response = await axios.get("/api/products");
                if (response.status === 200) {
                    setProducts(response.data.products);

                    console.log(response.data.msg);
                }
            } catch (err) {
                console.error(err);
            }
        };
        fetchProducts();
        return () => {};
    }, []);

    useEffect(() => {
        const fetchWishlist = async () => {
            try {
                if (uid) {
                    const response = await axios.get(`/api/wishlist/${uid}`);
                    if (response.status === 200) {
                        const newWishlist: Wishlist[] =
                            response.data.wishlist.map((item: any) => {
                                const { id, product_id, ...productProps } =
                                    item;

                                return {
                                    id,
                                    product: {
                                        id: product_id,
                                        ...productProps,
                                    },
                                };
                            });
                        setWishlist(newWishlist);
                        console.log(response.data.msg);
                    }
                }
            } catch (err) {
                console.error(err);
            }
        };
        fetchWishlist();
        return () => {};
    }, [uid]);

    useEffect(() => {
        const getProductIdsByUserId = (userId: string): number[] => {
            const userProductData = recommendations.find(
                (user) => user.user_id === userId
            );
            return userProductData ? userProductData.products : [];
        };
        if (userData && !loading) {
            setUserRecommendations(getProductIdsByUserId(userData.id));
        }
        return () => {};
    }, [userData, loading]);
    return (
        <Layout>
            <NavigationBar />
            <div className="home">
                <div className="home__hero">
                    <div className="home__hero__overlap">
                        <p className="home__hero__overlap__upto">
                            <span className="home__hero__overlap__upto__wrapper">
                                Up to 55% OFF
                                <br />
                            </span>
                            <span
                                className="home__hero__overlap__upto__wrapper"
                                style={{ fontSize: "28px" }}
                            >
                                with the new devices
                            </span>
                        </p>
                        <button
                            type="button"
                            className="home__hero__overlap__button"
                            onClick={() => navigate("/")}
                        >
                            BUY NOW <IoArrowForward />
                        </button>
                    </div>
                    <img
                        className="home__hero__overlap__iphone"
                        alt="Iphone"
                        src={iphone}
                    />
                </div>

                <div className="home__product">
                    <div className="home__product__header">
                        <h3 className="home__product__header__title">
                            All products
                        </h3>
                        <div>
                            <Link
                                to="/shops"
                                style={{
                                    textDecoration: "none",
                                    fontSize: "20px",
                                }}
                            >
                                View all <IoArrowForward />
                            </Link>
                        </div>
                    </div>
                    <div className="home__product__menu">
                        {products
                            .filter((product) => {
                                if (uid && userRecommendations.length > 0) {
                                    return (
                                        userRecommendations.includes(
                                            product.id
                                        ) && product.stock > 0
                                    );
                                } else return product.stock > 0;
                            })
                            .slice(0, DISPLAYED_NUMBER)
                            .map((product) => (
                                <ProductItem
                                    key={product.id}
                                    product={product}
                                    uid={uid}
                                    isWishlist={wishlist.some(
                                        (item) => item.product.id === product.id
                                    )}
                                    onAddingWishlist={() =>
                                        handleAddingWishlist(uid, product.id)
                                    }
                                    onAddingCart={() =>
                                        handleAddingCart(uid, product.id)
                                    }
                                />
                            ))}
                    </div>
                </div>
            </div>
        </Layout>
    );
};

export default HomePage;
