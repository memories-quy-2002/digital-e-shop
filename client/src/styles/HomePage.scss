.home {
    &__hero {
        background-color: #01b9e2;
        position: relative;
        width: 90%;
        height: 500px;
        margin: 2rem auto 8rem auto;
        &__overlap {
            height: 100%;
            left: 83px;
            position: absolute;
            width: 486px;
            &__upto {
                color: white;
                font-size: 54px;
                font-weight: 400;
                height: 100%;
                position: absolute;
                top: 25%;
                letter-spacing: 0;
                line-height: normal;
                &__wrapper {
                    color: white;
                }
            }
            &__button {
                background-color: black;
                color: #ffffff;
                font-size: 20px;
                font-weight: 400;
                height: 65px;
                left: 1px;
                position: absolute;
                text-align: center;
                top: 347px;
                width: 251px;
            }
            &__iphone {
                height: 100%;
                left: 1163px;
                object-fit: cover;
                position: absolute;
                top: 0;
                width: 372px;
                mix-blend-mode: multiply;
            }
        }
    }
    &__product {
        width: 90%;
        margin: 4rem auto;
        &__header {
            display: flex;
            justify-content: space-between;
            &__title {
                font-weight: bold;
                margin-bottom: 2rem;
            }
        }

        &__menu {
            width: 100%;
            display: grid;
            grid-template-columns: repeat(4, 1fr);
            gap: 20px;
            &__item {
                position: relative;
                padding-top: 1rem;
                height: fit-content;
                box-shadow: 0 4px 10px 3px rgba(0, 0, 0, 0.1882352941);
                &__image {
                    display: flex;
                    justify-content: center;
                    align-items: center;
                    height: 200px;
                    margin-bottom: 1rem;
                    padding: 1rem;
                    cursor: pointer;
                    img {
                        display: inline-block;
                        text-align: center;
                        height: 100%;
                        object-fit: contain;
                    }
                }
                &__wishlist {
                    position: absolute;
                    top: 10px;
                    right: 10px;
                    width: 40px;
                    height: 30px;
                    display: flex;
                    justify-content: center;
                    align-items: center;
                    background-color: #347988;
                    border-radius: 20px;
                    color: white;
                    opacity: 0;
                    transition: 0.2s opacity ease-in-out;
                }
                &__like {
                    position: absolute;
                    top: 10px;
                    right: 10px;
                    width: 40px;
                    aspect-ratio: 1;
                    display: flex;
                    justify-content: center;
                    align-items: center;
                    cursor: pointer;
                    opacity: 0;
                    transition: 0.2s opacity ease-in-out;
                }
                button {
                    background-color: white;
                    color: #0047ff;
                    width: 113px;
                    height: 44px;
                    border: 1px solid #0047ff;
                    font-weight: bold;
                    border-radius: 20px;
                    transition: 0.3s all ease-in-out;
                    &:hover {
                        background-color: #0047ff;
                        color: white;
                    }
                }
                &:hover &__like {
                    opacity: 1;
                }
                &:hover &__wishlist {
                    opacity: 1;
                }
            }
        }
    }
}
