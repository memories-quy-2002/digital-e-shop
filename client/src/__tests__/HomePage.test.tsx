import { fireEvent, render, screen } from "@testing-library/react";
import {
    BrowserRouter,
    MemoryRouter,
    Route,
    Routes,
    useLocation,
} from "react-router-dom";
import HomePage from "../components/pages/HomePage";
import ShopsPage from "../components/pages/ShopsPage";
import ToastProvider from "../context/ToastContext";

const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
    useNavigate: () => mockedUsedNavigate,
}));

const LocationDisplay = () => {
    const location = useLocation();
    return <div data-testid="location-display">{location.pathname}</div>;
};

describe("HomePage", () => {
    it("matches the HomePage snapshot", () => {
        const { asFragment } = render(
            <ToastProvider>
                <BrowserRouter>
                    <HomePage />
                </BrowserRouter>
            </ToastProvider>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("navigates to /shops when 'View all' is clicked", () => {
        render(
            <ToastProvider>
                <MemoryRouter initialEntries={["/"]}>
                    <Routes>
                        <Route path="/" element={<HomePage />} />
                        <Route path="/shops" element={<ShopsPage />} />
                    </Routes>
                    <LocationDisplay />
                </MemoryRouter>
            </ToastProvider>
        );

        const viewAllLink = screen.getByRole("link", { name: "View all" });
        expect(viewAllLink).toHaveAttribute("href", "/shops");
        if (viewAllLink) {
            fireEvent.click(viewAllLink);
        } else {
            throw new Error("Link 'View all' not found");
        }
        const location = screen.getByTestId("location-display");
        expect(location).toHaveTextContent("/shops");
    });
});
