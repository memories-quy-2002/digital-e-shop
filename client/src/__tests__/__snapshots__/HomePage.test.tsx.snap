// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage matches the HomePage snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="header__container"
    >
      <div
        class="header__container__info"
      >
        <div
          class="header__container__info__personal"
        >
          <div
            class="header__container__info__personal__item"
          >
            <svg
              color="white"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: white;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M391 480c-19.52 0-46.94-7.06-88-30-49.93-28-88.55-53.85-138.21-103.38C116.91 298.77 93.61 267.79 61 208.45c-36.84-67-30.56-102.12-23.54-117.13C45.82 73.38 58.16 62.65 74.11 52a176.3 176.3 0 0 1 28.64-15.2c1-.43 1.93-.84 2.76-1.21 4.95-2.23 12.45-5.6 21.95-2 6.34 2.38 12 7.25 20.86 16 18.17 17.92 43 57.83 52.16 77.43 6.15 13.21 10.22 21.93 10.23 31.71 0 11.45-5.76 20.28-12.75 29.81-1.31 1.79-2.61 3.5-3.87 5.16-7.61 10-9.28 12.89-8.18 18.05 2.23 10.37 18.86 41.24 46.19 68.51s57.31 42.85 67.72 45.07c5.38 1.15 8.33-.59 18.65-8.47 1.48-1.13 3-2.3 4.59-3.47 10.66-7.93 19.08-13.54 30.26-13.54h.06c9.73 0 18.06 4.22 31.86 11.18 18 9.08 59.11 33.59 77.14 51.78 8.77 8.84 13.66 14.48 16.05 20.81 3.6 9.53.21 17-2 22-.37.83-.78 1.74-1.21 2.75a176.49 176.49 0 0 1-15.29 28.58c-10.63 15.9-21.4 28.21-39.38 36.58A67.42 67.42 0 0 1 391 480z"
              />
            </svg>
            (+84) 123 456 7890
          </div>
          <div
            class="header__container__info__personal__item"
          >
            <svg
              color="white"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: white;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M464 80H48a16 16 0 0 0-16 16v320a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V96a16 16 0 0 0-16-16zM265.82 284.63a16 16 0 0 1-19.64 0L89.55 162.81l19.64-25.26L256 251.73l146.81-114.18 19.64 25.26z"
              />
            </svg>
            digital-e@gmail.com
          </div>
          <div
            class="header__container__info__personal__item"
          >
            <svg
              color="white"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: white;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M261.56 101.28a8 8 0 0 0-11.06 0L66.4 277.15a8 8 0 0 0-2.47 5.79L63.9 448a32 32 0 0 0 32 32H192a16 16 0 0 0 16-16V328a8 8 0 0 1 8-8h80a8 8 0 0 1 8 8v136a16 16 0 0 0 16 16h96.06a32 32 0 0 0 32-32V282.94a8 8 0 0 0-2.47-5.79z"
              />
              <path
                d="m490.91 244.15-74.8-71.56V64a16 16 0 0 0-16-16h-48a16 16 0 0 0-16 16v32l-57.92-55.38C272.77 35.14 264.71 32 256 32c-8.68 0-16.72 3.14-22.14 8.63l-212.7 203.5c-6.22 6-7 15.87-1.34 22.37A16 16 0 0 0 43 267.56L250.5 69.28a8 8 0 0 1 11.06 0l207.52 198.28a16 16 0 0 0 22.59-.44c6.14-6.36 5.63-16.86-.76-22.97z"
              />
            </svg>
            123 ABC Street, HCM City
          </div>
        </div>
        <div
          class="header__container__info__auth"
        >
          <strong>
            Welcome Anonymous
          </strong>
          <div
            class="header__container__info__auth__button"
          >
            <div>
              <a
                href="/login"
              >
                Login
              </a>
            </div>
            <div>
              |
            </div>
            <div>
              <a
                href="/signup"
              >
                Signup
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="header__container__main"
      >
        <div
          class="header__container__main__brand"
        >
          <a
            class="header__container__main__brand__link navbar-brand"
            href="/"
          >
            DIGITAL-E
          </a>
        </div>
        <div
          class="header__container__main__search"
        >
          <input
            class="header__container__main__search__bar"
            id=""
            name=""
            placeholder="Search product..."
            type="text"
          />
          <button>
            Search
          </button>
        </div>
        <div
          class="header__container__main__group"
        >
          <div
            class="header__container__main__group__item"
          >
            <svg
              fill="currentColor"
              height="28"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 448a32 32 0 0 1-18-5.57c-78.59-53.35-112.62-89.93-131.39-112.8-40-48.75-59.15-98.8-58.61-153C48.63 114.52 98.46 64 159.08 64c44.08 0 74.61 24.83 92.39 45.51a6 6 0 0 0 9.06 0C278.31 88.81 308.84 64 352.92 64c60.62 0 110.45 50.52 111.08 112.64.54 54.21-18.63 104.26-58.61 153-18.77 22.87-52.8 59.45-131.39 112.8a32 32 0 0 1-18 5.56z"
              />
            </svg>
            Wishlist
          </div>
          <div
            class="header__container__main__group__item"
          >
            <svg
              fill="currentColor"
              height="28"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="176"
                cy="416"
                r="32"
              />
              <circle
                cx="400"
                cy="416"
                r="32"
              />
              <path
                d="M456.8 120.78a23.92 23.92 0 0 0-18.56-8.78H133.89l-6.13-34.78A16 16 0 0 0 112 64H48a16 16 0 0 0 0 32h50.58l45.66 258.78A16 16 0 0 0 160 368h256a16 16 0 0 0 0-32H173.42l-5.64-32h241.66A24.07 24.07 0 0 0 433 284.71l28.8-144a24 24 0 0 0-5-19.93z"
              />
            </svg>
            Shopping Cart
          </div>
        </div>
      </div>
    </div>
    <nav
      class="navbar navbar-expand navbar-light bg-gray"
    >
      <div
        class="container"
      >
        <div
          class="navbar navbar-nav"
        >
          <a
            class="navbar__item nav-link"
            data-rr-ui-event-key="/"
            href="/"
          >
            Home
          </a>
          <a
            class="navbar__item nav-link"
            data-rr-ui-event-key="/shops"
            href="/shops"
          >
            shops
          </a>
          <a
            class="navbar__item nav-link"
            data-rr-ui-event-key="/hot-deals"
            href="/hot-deals"
          >
            hot deals
          </a>
          <a
            class="navbar__item nav-link"
            data-rr-ui-event-key="/about-us"
            href="/about-us"
          >
            about us
          </a>
          <a
            class="navbar__item nav-link"
            data-rr-ui-event-key="/contact-us"
            href="/contact-us"
          >
            contact us
          </a>
        </div>
      </div>
    </nav>
    <div
      class="home"
    >
      <div
        class="home__hero"
      >
        <div
          class="home__hero__overlap"
        >
          <p
            class="home__hero__overlap__upto"
          >
            <span
              class="home__hero__overlap__upto__wrapper"
            >
              Up to 55% OFF
              <br />
            </span>
            <span
              class="home__hero__overlap__upto__wrapper"
              style="font-size: 28px;"
            >
              with the new devices
            </span>
          </p>
          <button
            class="home__hero__overlap__button"
            type="button"
          >
            BUY NOW 
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m268 112 144 144-144 144m124-144H100"
                fill="none"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="48"
              />
            </svg>
          </button>
        </div>
        <img
          alt="Iphone"
          class="home__hero__overlap__iphone"
          src="iphone.jpg"
        />
      </div>
      <div
        class="home__product"
      >
        <div
          class="home__product__header"
        >
          <h3
            class="home__product__header__title"
          >
            All products
          </h3>
          <div>
            <a
              href="/shops"
              style="text-decoration: none; font-size: 20px;"
            >
              View all 
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m268 112 144 144-144 144m124-144H100"
                  fill="none"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="48"
                />
              </svg>
            </a>
          </div>
        </div>
        <div
          class="home__product__menu"
        />
      </div>
    </div>
    <div
      class="footer"
    >
      <div
        class="footer__about"
      >
        <h5
          style="font-weight: bold;"
        >
          About us
        </h5>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam.
        </p>
      </div>
      <div
        class="footer__menuGroup"
      >
        <div
          class="footer__menuGroup__menu"
        >
          <h5
            class="footer__menuGroup__menu__title"
          >
            Categories
          </h5>
          <div
            class="footer__menuGroup__menu__item"
          >
            Laptops
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Cameras
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Smartphones
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Accessories
          </div>
        </div>
        <div
          class="footer__menuGroup__menu"
        >
          <h5
            class="footer__menuGroup__menu__title"
          >
            Information
          </h5>
          <div
            class="footer__menuGroup__menu__item"
          >
            About us
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Contact us
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Privacy and Policy
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Terms and Condition
          </div>
        </div>
        <div
          class="footer__menuGroup__menu"
        >
          <h5
            class="footer__menuGroup__menu__title"
          >
            Service
          </h5>
          <div
            class="footer__menuGroup__menu__item"
          >
            My accounts
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Shopping cart
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Track my order
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Help
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="toast-container bottom-0 end-0 p-3"
    style="z-index: 1; position: fixed; bottom: 0px; right: 0px;"
  />
</DocumentFragment>
`;

exports[`matches the HomePage snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="header__container"
    >
      <div
        class="header__container__info"
      >
        <div
          class="header__container__info__personal"
        >
          <div
            class="header__container__info__personal__item"
          >
            <svg
              color="white"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: white;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M391 480c-19.52 0-46.94-7.06-88-30-49.93-28-88.55-53.85-138.21-103.38C116.91 298.77 93.61 267.79 61 208.45c-36.84-67-30.56-102.12-23.54-117.13C45.82 73.38 58.16 62.65 74.11 52a176.3 176.3 0 0 1 28.64-15.2c1-.43 1.93-.84 2.76-1.21 4.95-2.23 12.45-5.6 21.95-2 6.34 2.38 12 7.25 20.86 16 18.17 17.92 43 57.83 52.16 77.43 6.15 13.21 10.22 21.93 10.23 31.71 0 11.45-5.76 20.28-12.75 29.81-1.31 1.79-2.61 3.5-3.87 5.16-7.61 10-9.28 12.89-8.18 18.05 2.23 10.37 18.86 41.24 46.19 68.51s57.31 42.85 67.72 45.07c5.38 1.15 8.33-.59 18.65-8.47 1.48-1.13 3-2.3 4.59-3.47 10.66-7.93 19.08-13.54 30.26-13.54h.06c9.73 0 18.06 4.22 31.86 11.18 18 9.08 59.11 33.59 77.14 51.78 8.77 8.84 13.66 14.48 16.05 20.81 3.6 9.53.21 17-2 22-.37.83-.78 1.74-1.21 2.75a176.49 176.49 0 0 1-15.29 28.58c-10.63 15.9-21.4 28.21-39.38 36.58A67.42 67.42 0 0 1 391 480z"
              />
            </svg>
            (+84) 123 456 7890
          </div>
          <div
            class="header__container__info__personal__item"
          >
            <svg
              color="white"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: white;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M464 80H48a16 16 0 0 0-16 16v320a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V96a16 16 0 0 0-16-16zM265.82 284.63a16 16 0 0 1-19.64 0L89.55 162.81l19.64-25.26L256 251.73l146.81-114.18 19.64 25.26z"
              />
            </svg>
            digital-e@gmail.com
          </div>
          <div
            class="header__container__info__personal__item"
          >
            <svg
              color="white"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: white;"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M261.56 101.28a8 8 0 0 0-11.06 0L66.4 277.15a8 8 0 0 0-2.47 5.79L63.9 448a32 32 0 0 0 32 32H192a16 16 0 0 0 16-16V328a8 8 0 0 1 8-8h80a8 8 0 0 1 8 8v136a16 16 0 0 0 16 16h96.06a32 32 0 0 0 32-32V282.94a8 8 0 0 0-2.47-5.79z"
              />
              <path
                d="m490.91 244.15-74.8-71.56V64a16 16 0 0 0-16-16h-48a16 16 0 0 0-16 16v32l-57.92-55.38C272.77 35.14 264.71 32 256 32c-8.68 0-16.72 3.14-22.14 8.63l-212.7 203.5c-6.22 6-7 15.87-1.34 22.37A16 16 0 0 0 43 267.56L250.5 69.28a8 8 0 0 1 11.06 0l207.52 198.28a16 16 0 0 0 22.59-.44c6.14-6.36 5.63-16.86-.76-22.97z"
              />
            </svg>
            123 ABC Street, HCM City
          </div>
        </div>
        <div
          class="header__container__info__auth"
        >
          <strong>
            Welcome Anonymous
          </strong>
          <div
            class="header__container__info__auth__button"
          >
            <div>
              <a
                href="/login"
              >
                Login
              </a>
            </div>
            <div>
              |
            </div>
            <div>
              <a
                href="/signup"
              >
                Signup
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="header__container__main"
      >
        <div
          class="header__container__main__brand"
        >
          <a
            class="header__container__main__brand__link navbar-brand"
            href="/"
          >
            DIGITAL-E
          </a>
        </div>
        <div
          class="header__container__main__search"
        >
          <input
            class="header__container__main__search__bar"
            id=""
            name=""
            placeholder="Search product..."
            type="text"
          />
          <button>
            Search
          </button>
        </div>
        <div
          class="header__container__main__group"
        >
          <div
            class="header__container__main__group__item"
          >
            <svg
              fill="currentColor"
              height="28"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 448a32 32 0 0 1-18-5.57c-78.59-53.35-112.62-89.93-131.39-112.8-40-48.75-59.15-98.8-58.61-153C48.63 114.52 98.46 64 159.08 64c44.08 0 74.61 24.83 92.39 45.51a6 6 0 0 0 9.06 0C278.31 88.81 308.84 64 352.92 64c60.62 0 110.45 50.52 111.08 112.64.54 54.21-18.63 104.26-58.61 153-18.77 22.87-52.8 59.45-131.39 112.8a32 32 0 0 1-18 5.56z"
              />
            </svg>
            Wishlist
          </div>
          <div
            class="header__container__main__group__item"
          >
            <svg
              fill="currentColor"
              height="28"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="28"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="176"
                cy="416"
                r="32"
              />
              <circle
                cx="400"
                cy="416"
                r="32"
              />
              <path
                d="M456.8 120.78a23.92 23.92 0 0 0-18.56-8.78H133.89l-6.13-34.78A16 16 0 0 0 112 64H48a16 16 0 0 0 0 32h50.58l45.66 258.78A16 16 0 0 0 160 368h256a16 16 0 0 0 0-32H173.42l-5.64-32h241.66A24.07 24.07 0 0 0 433 284.71l28.8-144a24 24 0 0 0-5-19.93z"
              />
            </svg>
            Shopping Cart
          </div>
        </div>
      </div>
    </div>
    <nav
      class="navbar navbar-expand navbar-light bg-gray"
    >
      <div
        class="container"
      >
        <div
          class="navbar navbar-nav"
        >
          <a
            class="navbar__item nav-link"
            data-rr-ui-event-key="/"
            href="/"
          >
            Home
          </a>
          <a
            class="navbar__item nav-link"
            data-rr-ui-event-key="/shops"
            href="/shops"
          >
            shops
          </a>
          <a
            class="navbar__item nav-link"
            data-rr-ui-event-key="/hot-deals"
            href="/hot-deals"
          >
            hot deals
          </a>
          <a
            class="navbar__item nav-link"
            data-rr-ui-event-key="/about-us"
            href="/about-us"
          >
            about us
          </a>
          <a
            class="navbar__item nav-link"
            data-rr-ui-event-key="/contact-us"
            href="/contact-us"
          >
            contact us
          </a>
        </div>
      </div>
    </nav>
    <div
      class="home"
    >
      <div
        class="home__hero"
      >
        <div
          class="home__hero__overlap"
        >
          <p
            class="home__hero__overlap__upto"
          >
            <span
              class="home__hero__overlap__upto__wrapper"
            >
              Up to 55% OFF
              <br />
            </span>
            <span
              class="home__hero__overlap__upto__wrapper"
              style="font-size: 28px;"
            >
              with the new devices
            </span>
          </p>
          <button
            class="home__hero__overlap__button"
            type="button"
          >
            BUY NOW 
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m268 112 144 144-144 144m124-144H100"
                fill="none"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="48"
              />
            </svg>
          </button>
        </div>
        <img
          alt="Iphone"
          class="home__hero__overlap__iphone"
          src="iphone.jpg"
        />
      </div>
      <div
        class="home__product"
      >
        <div
          class="home__product__header"
        >
          <h3
            class="home__product__header__title"
          >
            All products
          </h3>
          <div>
            <a
              href="/shops"
              style="text-decoration: none; font-size: 20px;"
            >
              View all 
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 512 512"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m268 112 144 144-144 144m124-144H100"
                  fill="none"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="48"
                />
              </svg>
            </a>
          </div>
        </div>
        <div
          class="home__product__menu"
        />
      </div>
    </div>
    <div
      class="footer"
    >
      <div
        class="footer__about"
      >
        <h5
          style="font-weight: bold;"
        >
          About us
        </h5>
        <p>
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam.
        </p>
      </div>
      <div
        class="footer__menuGroup"
      >
        <div
          class="footer__menuGroup__menu"
        >
          <h5
            class="footer__menuGroup__menu__title"
          >
            Categories
          </h5>
          <div
            class="footer__menuGroup__menu__item"
          >
            Laptops
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Cameras
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Smartphones
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Accessories
          </div>
        </div>
        <div
          class="footer__menuGroup__menu"
        >
          <h5
            class="footer__menuGroup__menu__title"
          >
            Information
          </h5>
          <div
            class="footer__menuGroup__menu__item"
          >
            About us
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Contact us
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Privacy and Policy
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Terms and Condition
          </div>
        </div>
        <div
          class="footer__menuGroup__menu"
        >
          <h5
            class="footer__menuGroup__menu__title"
          >
            Service
          </h5>
          <div
            class="footer__menuGroup__menu__item"
          >
            My accounts
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Shopping cart
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Track my order
          </div>
          <div
            class="footer__menuGroup__menu__item"
          >
            Help
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="toast-container bottom-0 end-0 p-3"
    style="z-index: 1; position: fixed; bottom: 0px; right: 0px;"
  />
</DocumentFragment>
`;
